qbft/round_change.go

```diff
@@ -0,0 +1,347 @@
+package qbft
+
+import (
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/pkg/errors"
+)
+
+func (i *Instance) uponRoundChange(
+	instanceStartValue []byte,
+	signedRoundChange *SignedMessage,
+	roundChangeMsgContainer *MsgContainer,
+	valCheck ProposedValueCheckF,
+) error {
+	if err := validRoundChange(i.State, i.config, signedRoundChange, i.State.Height, signedRoundChange.Message.Round); err != nil {
+		return errors.Wrap(err, "round change msg invalid")
+	}
+
+	addedMsg, err := roundChangeMsgContainer.AddFirstMsgForSignerAndRound(signedRoundChange)
+	if err != nil {
+		return errors.Wrap(err, "could not add round change msg to container")
+	}
+	if !addedMsg {
+		return nil // UponCommit was already called
+	}
+
+	justifiedRoundChangeMsg, err := hasReceivedProposalJustificationForLeadingRound(
```

Comment:
I'm not saying this code is incorrect, but what follow from this line until line 62 feels pretty confusing to me. at this level, I would expect a more clear way to tally the round-change messages received and a more visible check of who the leader is before broadcasting a proposal.